### PR TITLE
[ARM/CRT] Floating point control functions implementation

### PR DESCRIPTION
**Pull request by theR4K**
## TODO

- [ ] test

### CODE REVIEW

File: sdk/lib/crt/float/arm/fpscr.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/*
+ * PROJECT:     ReactOS CRT library
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     arm fpcsr description
```
Comment (binarymaster) [inline sdk/lib/crt/float/arm/fpscr.h:4]:
```suggestion
 * PURPOSE:     Headers for ARM fpcsr
```

File: sdk/lib/crt/float/arm/_controlfp.c [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    TRACE("Control word before : %08x\n", fpscr.raw);
+
+    /* Convert into mask constants */
+    if (!(fpscr.data.ex_control & ARM_CW_IM)) flags |= _EM_INVALID;
+    if (!(fpscr.data.ex_control & ARM_CW_ZM)) flags |= _EM_ZERODIVIDE;
+    if (!(fpscr.data.ex_control & ARM_CW_OM)) flags |= _EM_OVERFLOW;
+    if (!(fpscr.data.ex_control & ARM_CW_UM)) flags |= _EM_UNDERFLOW;
+    if (!(fpscr.data.ex_control & ARM_CW_PM)) flags |= _EM_INEXACT;
+    if (!(fpscr.data.ex_control & ARM_CW_DM)) flags |= _EM_DENORMAL;
+    switch(fpscr.data.rounding_mode) {
```
Comment (binarymaster) [inline sdk/lib/crt/float/arm/_controlfp.c:35]:
```suggestion

    switch (fpscr.data.rounding_mode)
    {
```

File: sdk/include/crt/float.h [hunk 3]
```diff
@@ -77,6 +77,14 @@
 #define	_FPCLASS_PINF	0x0200	/* Positive Infinity */
 #endif /* __MINGW_FPCLASS_DEFINED */
 
+/* _statusfp bit flags */
+#define _SW_INEXACT    0x00000001  /* inexact (precision) */
+#define _SW_UNDERFLOW  0x00000002  /* underflow */
+#define _SW_OVERFLOW   0x00000004  /* overflow */
+#define _SW_ZERODIVIDE 0x00000008  /* zero divide */
+#define _SW_INVALID    0x00000010  /* invalid */
+#define _SW_DENORMAL    0x00080000 /* denormal status bit */
```
Comment (binarymaster) [inline sdk/include/crt/float.h:86]:
```suggestion
#define _SW_DENORMAL   0x00080000  /* denormal status bit */
```

File: sdk/lib/crt/float/arm/__getfp.s [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/*
+ * PROJECT:     ReactOS CRT library
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     arm fpscr getter
```
Comment (binarymaster) [inline sdk/lib/crt/float/arm/__getfp.s:4]:
```suggestion
 * PURPOSE:     ARM fpscr getter
```

File: sdk/lib/crt/float/arm/__setfp.s [hunk 5]
```diff
@@ -0,0 +1,25 @@
+/*
+ * PROJECT:     ReactOS CRT library
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     arm fpscr setter
```
Comment (binarymaster) [inline sdk/lib/crt/float/arm/__setfp.s:4]:
```suggestion
 * PURPOSE:     ARM fpscr setter
```

File: sdk/lib/crt/float/arm/_clearfp.c [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    unsigned int status;
+
+    fpscr.raw = __getfp();
+    status = _statusfp();
+    
+    fpscr.data.exception = fpscr.data.exception & ~ARM_CW_STATUS_MASK;
+    
+    __setfp(fpscr.raw);
+    return status;
+}
```
Comment (binarymaster) [inline sdk/lib/crt/float/arm/_clearfp.c:22]: No newline at end of file

File: sdk/lib/crt/float/arm/_controlfp.c [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+    /* Convert (masked) value back to fp word */
+    fpscr.raw = 0;
+    if (!(flags & _EM_INVALID))    fpscr.data.ex_control |= ARM_CW_IM;
+    if (!(flags & _EM_ZERODIVIDE)) fpscr.data.ex_control |= ARM_CW_ZM;
+    if (!(flags & _EM_OVERFLOW))   fpscr.data.ex_control |= ARM_CW_OM;
+    if (!(flags & _EM_UNDERFLOW))  fpscr.data.ex_control |= ARM_CW_UM;
+    if (!(flags & _EM_INEXACT))    fpscr.data.ex_control |= ARM_CW_PM;
+    if (!(flags & _EM_DENORMAL))   fpscr.data.ex_control |= ARM_CW_DM;
+    switch(flags & (_RC_UP | _RC_DOWN)) {
```
Comment (binarymaster) [inline sdk/lib/crt/float/arm/_controlfp.c:52]:
```suggestion

    switch (flags & (_RC_UP | _RC_DOWN))
    {
```

File: sdk/lib/crt/float/arm/_controlfp.c [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        case _RC_UP|_RC_DOWN: fpscr.data.rounding_mode = ARM_CW_RC_ZERO; break;
+        case _RC_UP:          fpscr.data.rounding_mode = ARM_CW_RC_UP; break;
```
Comment (binarymaster) [inline sdk/lib/crt/float/arm/_controlfp.c:54]: If I understand correctly, `case _RC_UP` will never be triggered, as above one `case _RC_UP|_RC_DOWN` has a break.

File: sdk/lib/crt/float/arm/_fpreset.c [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+ * PURPOSE:     Implementation of _fpreset
+ * COPYRIGHT:   Copyright 2021 Roman Masanin <36927roma@gmail.com>
+ */
+
+#include "fpscr.h"
+
+void _fpreset(void)
+{
+    __setfp(0x0);
+}
```
Comment (binarymaster) [inline sdk/lib/crt/float/arm/_fpreset.c:13]: No newline at end of file

File: sdk/lib/crt/float/arm/fpscr.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        unsigned int stride: 2;
+        unsigned int rounding_mode: 2;
+        unsigned int flush_to_zero: 1;
+        unsigned int unused4: 3;
+        unsigned int status_flag: 4;
+    } data;
+} ARM_FPSCR;
+
+void __setfp(unsigned int);
+unsigned int __getfp(void);
```
Comment (binarymaster) [inline sdk/lib/crt/float/arm/fpscr.h:41]: No newline at end of file

File: sdk/lib/crt/float/arm/_controlfp.c (hunk 8 above)
Comment (theR4K) [reply sdk/lib/crt/float/arm/_controlfp.c:54]: no, this is a swich case and they work as ==.

**Review by binarymaster — APPROVED** [review]
Looks good to me, and it also fixes [CORE-17757](https://jira.reactos.org/browse/CORE-17757) for me. :+1:  Gonna merge without API tests.

**Pull request merged**